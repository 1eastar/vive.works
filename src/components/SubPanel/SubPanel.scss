@import '@styles/global.scss';

.TOC-container {
  padding: 22px 10px 0;
  height: 100%;

  .header {
    display: flex;
    flex-direction: row;
    align-items: center;
    justify-content: center;
    padding: 30px 10px;

    .title {
      color: var(--text-color-primary);
      font-weight: bold;
      font-size: $font-size-xl;
      margin-right: auto;
      cursor: pointer;
    }
    
    .icon {
      padding: 0 3px;

      svg {
        border-radius: 50%;
        transition: box-shadow 0.1s ease-in-out;
        cursor: pointer;

        &:hover {
          box-shadow: 0 1px 20px var(--black25);
        }

        @media (prefers-color-scheme: dark) {
          path {
            fill: var(--text-color-primary);
          }
        }
      }
    }
  }

  .TOC-item-title {
    font-weight: bold;
    font-size: $font-size-xxs;
    color: var(--text-color-primary-30);
    letter-spacing: 0.35em;
    user-select: none;
    padding: 10px;
  }
}