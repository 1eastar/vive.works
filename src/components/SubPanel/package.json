{
  "name": "SubPanel",
  "version": "0.0.0",
  "private": true,
  "main": "SubPanel.tsx"
}