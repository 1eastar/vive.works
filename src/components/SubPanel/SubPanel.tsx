/* External */
import { useCallback } from 'react'

/* Internal */
import { VERSION } from '@lib/version/version'
import { isBrowser } from '@utils/browser.utils'
import { EMAIL, GITHUB_REPOSITORY_URL, GITHUB_URL } from '@commons/constants/info.constant'
import { SUB_PANEL_TITLE } from '@commons/constants/subPanel.constant'
import TagCollection from './TagCollection'
import TOC from './TOC'
import * as styles from './SubPanel.scss'

import GithubIcon from '-!svg-react-loader!@statics/icons/github.svg'
import EmailIcon from '-!svg-react-loader!@statics/icons/email.svg'

function SubPanel() {
  const pathname = isBrowser ? window.location.pathname : ''

  const onClickHeader = useCallback(() => {
    window.open(GITHUB_REPOSITORY_URL, '_blank', 'noopener noreferrer')
  }, [])

  const onClickGithubIcon = useCallback(() => {
    window.open(GITHUB_URL, '_blank', 'noopener noreferrer')
  }, [])

  const onClickEmailIcon = useCallback(() => {
    window.location.href = `mailto:${EMAIL}`
  }, [])

  return (
    <div className={styles.TOCContainer}>
      <div className={styles.header}>
        <div
          className={styles.title}
          onClick={onClickHeader}
        >
          Vive works v{VERSION}
        </div>
        <div className={styles.icon} onClick={onClickGithubIcon}>
          <GithubIcon />
        </div>
        <div className={styles.icon} onClick={onClickEmailIcon}>
          <EmailIcon />
        </div>
      </div>
      <div className={styles.TOCItemTitle}>
        { pathname === '/'
          ? SUB_PANEL_TITLE['Main']
          : SUB_PANEL_TITLE['Post']
        }
      </div>
      { pathname === '/'
        ? <TagCollection />
        : <TOC />
      }
    </div>
  )
}

export default SubPanel
