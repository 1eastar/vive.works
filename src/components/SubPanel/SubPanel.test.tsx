/* External */
import { render, cleanup, fireEvent } from '@testing-library/react'

/* Internal */
import SubPanel from './SubPanel'

beforeAll(() => {
  const mockIntersectionObserver = jest.fn().mockReturnValue({
    observe: () => null,
    unobserve: () => null,
    disconnect: () => null
  })

  window.IntersectionObserver = mockIntersectionObserver
})

afterEach(cleanup)

describe('<SubPanel/>', () => {
  describe('게시글 페이지', () => {
    test('[snapshot] 게시글 페이지에서는 헤더, 아이콘들과 함께 TOC를 렌더링한다.', () => {
      const { SubPanelEl } = renderSubPanel()
  
      expect(SubPanelEl()).toMatchSnapshot()
    })

    test('헤더 클릭 시 메인 화면으로 이동한다.', () => {
      const { clickHeader } = renderSubPanel()
  
      expect(location.href).toBe('https://vive.works/mockPathname?tag=mockTag&key1=value1&key2=value2')
      clickHeader()
      expect(location.href).toBe('https://vive.works/')
    })
  })

  describe('메인 페이지', () => {
    test('[snapshot] 헤더, 아이콘들과 함께 Tag collection을 렌더링한다.', () => {
      const { SubPanelEl } = renderSubPanel()
  
      expect(SubPanelEl()).toMatchSnapshot()
    })

    test('헤더 클릭 시 메인 화면으로 이동한다.', () => {
      const { clickHeader } = renderSubPanel()
  
      expect(location.href).toBe('https://vive.works/')
      clickHeader()
      expect(location.href).toBe('https://vive.works/')
    })
  })
})


/* Helper function */
function renderSubPanel() {
  const location = { pathname: '/' } as Location
  const Result = render(<SubPanel location={location}/>)

  const SubPanelEl = () => Result.getByTestId('SubPanel-container')

  const Header_Main = () => Result.getByText('Vive works')

  function clickHeader() {
    fireEvent.click(Header_Main())
  }


  return {
    SubPanelEl,
    clickHeader,
  }
}
