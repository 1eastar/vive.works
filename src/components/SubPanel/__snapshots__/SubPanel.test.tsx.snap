// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<SubPanel/> 게시글 페이지 [snapshot] 게시글 페이지에서는 헤더, 아이콘들과 함께 TOC를 렌더링한다. 1`] = `
<div
  data-testid="SubPanel-container"
>
  <div>
    <div>
      Vive works
    </div>
    <div>
      <test-file-stub />
    </div>
    <div>
      <test-file-stub />
    </div>
  </div>
  <div>
    Tags
  </div>
  <div
    data-testid="tagCollection"
  >
    <div
      class=""
      data-testid="tag-All"
    >
      # 
      All
         (
      5
      )
    </div>
    <div
      class=""
      data-testid="tag-tag1"
    >
      # 
      tag1
         (
      1
      )
    </div>
    <div
      class=""
      data-testid="tag-tag2"
    >
      # 
      tag2
         (
      3
      )
    </div>
    <div
      class=""
      data-testid="tag-tag3"
    >
      # 
      tag3
         (
      2
      )
    </div>
    <div
      class=""
      data-testid="tag-tag4"
    >
      # 
      tag4
         (
      1
      )
    </div>
    <div
      class=""
      data-testid="tag-tag5"
    >
      # 
      tag5
         (
      1
      )
    </div>
  </div>
</div>
`;

exports[`<SubPanel/> 메인 페이지 [snapshot] 헤더, 아이콘들과 함께 Tag collection을 렌더링한다. 1`] = `
<div
  data-testid="SubPanel-container"
>
  <div>
    <div>
      Vive works
    </div>
    <div>
      <test-file-stub />
    </div>
    <div>
      <test-file-stub />
    </div>
  </div>
  <div>
    Tags
  </div>
  <div
    data-testid="tagCollection"
  >
    <div
      class=""
      data-testid="tag-All"
    >
      # 
      All
         (
      5
      )
    </div>
    <div
      class=""
      data-testid="tag-tag1"
    >
      # 
      tag1
         (
      1
      )
    </div>
    <div
      class=""
      data-testid="tag-tag2"
    >
      # 
      tag2
         (
      3
      )
    </div>
    <div
      class=""
      data-testid="tag-tag3"
    >
      # 
      tag3
         (
      2
      )
    </div>
    <div
      class=""
      data-testid="tag-tag4"
    >
      # 
      tag4
         (
      1
      )
    </div>
    <div
      class=""
      data-testid="tag-tag5"
    >
      # 
      tag5
         (
      1
      )
    </div>
  </div>
</div>
`;
